src/ln/channel.rs

```diff
@@ -1820,13 +1853,48 @@ impl Channel {
 						update_fulfill_htlcs: Vec::new(),
 						update_fail_htlcs,
 						update_fail_malformed_htlcs,
+						update_fee: update_fee_msg,
 						commitment_signed
 					}), to_forward_infos, revoked_htlcs, monitor_update))
 				} else {
 					Ok((None, to_forward_infos, revoked_htlcs, self.channel_monitor.clone()))
 				}
 			}
 		}
+
+	}
+
+	pub fn send_update_fee(&mut self, feerate_per_kw: u64) -> Result<Option<msgs::UpdateFee>, HandleError> {
+		if !self.channel_outbound {
+			return Err(HandleError{err: "Cannot send fee from inbound channel", action: None});
+		}
+
+		if (self.channel_state & (ChannelState::ChannelFunded as u32 | BOTH_SIDES_SHUTDOWN_MASK)) != (ChannelState::ChannelFunded as u32) {
+			return Err(HandleError{err: "Cannot update fee until channel is fully established and we haven't started shutting down", action: None});
+		}
+
+		if (self.channel_state & (ChannelState::AwaitingRemoteRevoke as u32)) == (ChannelState::AwaitingRemoteRevoke as u32) {
+			self.holding_cell_update_fee = Some(feerate_per_kw);
+			return Ok(None);
+		}
+
+		debug_assert!(self.pending_update_fee.is_none());
+		self.pending_update_fee = Some(feerate_per_kw);
+
+		Ok(Some(msgs::UpdateFee {
+			channel_id: self.channel_id,
+			feerate_per_kw: feerate_per_kw as u32,
+		}))
+	}
+
+	pub fn send_update_fee_and_commit(&mut self, feerate_per_kw: u64) -> Result<Option<(msgs::UpdateFee, msgs::CommitmentSigned, ChannelMonitor)>, HandleError> {
+		match self.send_update_fee(feerate_per_kw)? {
+			Some(update_fee) => {
+				let (commitment_signed, monitor_update) = self.send_commitment_no_status_check()?;
```

Comment:
Hmm, I don't think this is quite right:
When we call send_update_fee_and_commit we'll create the update_fee message, send it to our peer, then create a commitment_signed message with the new fee (that is all correct). However, when we create the commitment_signed message we'll set self.feerate_per_kw, which means we'll be incorrect in some race conditions.

What should happen is this:
```
Us                                    remote
update_fee                       ->
commitment_signed (incl new fee) ->
                                 <- update_add_htlc
                                 <- commitment_signed (not incl new fee)
```
But we'll expect their commitment_signed to use the new feerate. This is incorrect and we should instead only expect that after they send a new revoke_and_ack.

You may want to add a test which tests the above message exchanges.

Reply:
Not sure how to generate the condition above with our code. The handler of commitment_signed generates revoke_and_ack and commitment_signed, and put the channel in AwaitingRemoteRevoke state. So the second commitment signed will use new fee after 'remote' having received revoke_and_ack from 'Us.' 

If you meant update_add_htlc/commitment_signed arriving to 'Us' when the first commitment_signed in transit, then it's covered, I think, because 'Us' does not update self.feerate until it gets revoke_and_ack from 'remote.'

Reply:
You should be able to create two channelmanagers - one call update_fee and get the three messages out of it, on the other send_payment. Now forward the messages to both ends but only after calling both and ensure they all behave correctly.